### ISSUE TITLE
Allow JSDoc make arrow parens needed

### ISSUE THREAD

**Issue by tiansh** [eslint/eslint#12995]
**What rule do you want to change?**

    arrow-parens: "as-needed"

**Does this change cause the rule to produce more or fewer warnings?**

Fewer

**How will the change be implemented? (New option, new default behavior, etc.)?**

Prefer new option, make default should be fine too.

**Please provide some example code that this change will affect:**

```js
/*eslint arrow-parens: ["error", "as-needed"]*/

const f = (/** @type {number} */a) => a + a;
f(42);
```

[Demo](https://eslint.org/demo#eyJ0ZXh0IjoiLyplc2xpbnQgYXJyb3ctcGFyZW5zOiBbXCJlcnJvclwiLCBcImFzLW5lZWRlZFwiXSovXG5cbmNvbnN0IGYgPSAoLyoqIEB0eXBlIHtudW1iZXJ9ICovYSkgPT4gYSArIGE7XG5mKDQyKTtcbiIsIm9wdGlvbnMiOnsicGFyc2VyT3B0aW9ucyI6eyJlY21hVmVyc2lvbiI6MTEsInNvdXJjZVR5cGUiOiJzY3JpcHQiLCJlY21hRmVhdHVyZXMiOnt9fSwicnVsZXMiOnsiY29uc3RydWN0b3Itc3VwZXIiOjIsImZvci1kaXJlY3Rpb24iOjIsImdldHRlci1yZXR1cm4iOjIsIm5vLWFzeW5jLXByb21pc2UtZXhlY3V0b3IiOjIsIm5vLWNhc2UtZGVjbGFyYXRpb25zIjoyLCJuby1jbGFzcy1hc3NpZ24iOjIsIm5vLWNvbXBhcmUtbmVnLXplcm8iOjIsIm5vLWNvbmQtYXNzaWduIjoyLCJuby1jb25zdC1hc3NpZ24iOjIsIm5vLWNvbnN0YW50LWNvbmRpdGlvbiI6Miwibm8tY29udHJvbC1yZWdleCI6Miwibm8tZGVidWdnZXIiOjIsIm5vLWRlbGV0ZS12YXIiOjIsIm5vLWR1cGUtYXJncyI6Miwibm8tZHVwZS1jbGFzcy1tZW1iZXJzIjoyLCJuby1kdXBlLWtleXMiOjIsIm5vLWR1cGxpY2F0ZS1jYXNlIjoyLCJuby1lbXB0eSI6Miwibm8tZW1wdHktY2hhcmFjdGVyLWNsYXNzIjoyLCJuby1lbXB0eS1wYXR0ZXJuIjoyLCJuby1leC1hc3NpZ24iOjIsIm5vLWV4dHJhLWJvb2xlYW4tY2FzdCI6Miwibm8tZXh0cmEtc2VtaSI6Miwibm8tZmFsbHRocm91Z2giOjIsIm5vLWZ1bmMtYXNzaWduIjoyLCJuby1nbG9iYWwtYXNzaWduIjoyLCJuby1pbm5lci1kZWNsYXJhdGlvbnMiOjIsIm5vLWludmFsaWQtcmVnZXhwIjoyLCJuby1pcnJlZ3VsYXItd2hpdGVzcGFjZSI6Miwibm8tbWlzbGVhZGluZy1jaGFyYWN0ZXItY2xhc3MiOjIsIm5vLW1peGVkLXNwYWNlcy1hbmQtdGFicyI6Miwibm8tbmV3LXN5bWJvbCI6Miwibm8tb2JqLWNhbGxzIjoyLCJuby1vY3RhbCI6Miwibm8tcHJvdG90eXBlLWJ1aWx0aW5zIjoyLCJuby1yZWRlY2xhcmUiOjIsIm5vLXJlZ2V4LXNwYWNlcyI6Miwibm8tc2VsZi1hc3NpZ24iOjIsIm5vLXNoYWRvdy1yZXN0cmljdGVkLW5hbWVzIjoyLCJuby1zcGFyc2UtYXJyYXlzIjoyLCJuby10aGlzLWJlZm9yZS1zdXBlciI6Miwibm8tdW5kZWYiOjIsIm5vLXVuZXhwZWN0ZWQtbXVsdGlsaW5lIjoyLCJuby11bnJlYWNoYWJsZSI6Miwibm8tdW5zYWZlLWZpbmFsbHkiOjIsIm5vLXVuc2FmZS1uZWdhdGlvbiI6Miwibm8tdW51c2VkLWxhYmVscyI6Miwibm8tdW51c2VkLXZhcnMiOjIsIm5vLXVzZWxlc3MtY2F0Y2giOjIsIm5vLXVzZWxlc3MtZXNjYXBlIjoyLCJuby13aXRoIjoyLCJyZXF1aXJlLXlpZWxkIjoyLCJ1c2UtaXNuYW4iOjIsInZhbGlkLXR5cGVvZiI6Mn0sImVudiI6e319fQ==)

**What does the rule currently do for this code?**

    3:33 - Unexpected parentheses around single function argument.

**What will the rule do after it's changed?**

No error.

**Are you willing to submit a pull request to implement this change?**

Not.

----

My IDE seems do not know type of variable `a` in `(/** @type {number} */a) => a` if the braces are omitted. I want to make JSDoc a exception of validating `arrow-parens: "as-needed"`.

**Comment by mdjermanovic** [comment]
I'd support an option to allow existing parentheses if there are any comments inside.

**Comment by anikethsaha** [comment]
I would like to work on this. 

Here is what I am planning as @mdjermanovic suggested to have an option for this.
options name : `requireCommentsParams`
It will check for comments in params. It will warn only when 
- no comments are passed
- When empty comments are passed

Example:

```js
/*eslint arrow-parens: [2, "as-needed", { "requireCommentsParams": true}]*/

const f = (/** @type {number} */a) => a + a;   // correct
const f = (/* */a) => a + a;    // Incorrect
const f = (a) => a + a;    // Incorrect
```

Though I would even back this without having a separate option for it but having a separate option is good for initial stages then move to default

[Rough Implementation](https://astexplorer.net/#/gist/291a0da4a4c14395ee50d6d20cd5120b/78582a9302150c6f992b0a26aa62af5a97d6e2f4)

**Comment by kaicataldo** [comment]
@anikethsaha Thanks for being willing to work on this. Please note that this proposal hasn't been accepted yet, and we generally recommend that you wait until an issue has been accepted before implementing (to avoid throwaway work).

Regarding handling comments, ESLint doesn't use `(leading|trailing)Comments` any more (this behavior was [dropped in v4](https://eslint.org/docs/user-guide/migrating-to-4.0.0#ast-nodes-no-longer-have-comment-properties). ESLint now handles comments in the context of tokens, rather than the AST. In addition to the AST, Espree also returns a top-level `tokens` and `comments` key, and ESLint uses these arrays to create a token store that powers the the token-iteration APIs. Please check out the docs [here](https://eslint.org/docs/developer-guide/working-with-rules#accessing-comments) on more information for handling comments.

**Comment by anikethsaha** [comment]
> @anikethsaha Thanks for being willing to work on this. Please note that this proposal hasn't been accepted yet, and we generally recommend that you wait until an issue has been accepted before implementing (to avoid throwaway work).

Sure ! I will wait till it gets accepted. 


> Please check out the docs here on more information for handling comments.

Thanks for the link !
I Will check that 👍

**Comment by mdjermanovic** [comment]
I think there is no need to check if the comment is not empty, just whether there is at least one comment between `(` and  `)`, so all of the following would be allowed:

```js
const f = (/**/a) => a + a;

const g = (//
    a) => a + a;

const h = (a/**/) => a + a;

const i = (a/**/,) => a + a;

const j = (a //
    ) => a + a;
```

**Comment by anikethsaha** [comment]
> ```js
> 
> const g = (//
>     a) => a + a;
>
> const j = (a //
>     ) => a + a;
> ```

I see, these cases do make sense to keep it as it is 👍 

 Lost of comments is not so great options especially for use cases like jsdoc
I would back this enhancement as it seems required

**Comment by mdjermanovic** [comment]
I'll champion this.

**Comment by kaicataldo** [comment]
@mdjermanovic Are you still championing this? If so, I think we need to try to reach consensus on its approval. It's been open for quite a while now.

**Comment by mysticatea** [comment]
It looks like a bug to me. Autofix should not remove comments accidentally.

**Comment by mdjermanovic** [comment]
There is indeed the autofix bug, and there is this request to not warn at all about unneeded parentheses when there's a comment inside (if the new option is set or as the new default behavior).

Is it accepted to not warn at all, as the new default behavior (no option) which would solve both the bug and this issue?

**Comment by anikethsaha** [comment]
I think it should be default behavior cause removing of comments seems bug right?

**Comment by golopot** [comment]
Aside from the autofix bug, there is the problem that the parentheses are needed when the code is type checked by [TypeScript --checkjs](https://www.typescriptlang.org/docs/handbook/type-checking-javascript-files.html#type).

```js
// The type annotation applies to parameter x.
const f = (/** @type {string} */ x) => x

// The type annotation applies to the arrow function expression.
const f = /** @type {string} */ x => x
```

**Comment by mdjermanovic** [comment]
> I think it should be default behavior cause removing of comments seems bug right?

Yes, but the bug can be fixed by modifying the fixer to remove just parens and keep everything inside, or more likely by not doing the autofix if there are any comments inside (given the @golopot's comment).

This is different from not reporting the error at all. Which is also fine, just wanted to doublecheck what's exactly accepted.

**Comment by anikethsaha** [comment]
ahh, yeah IMO, not to warn should be the default behavior.

**Comment by mysticatea** [comment]
@mdjermanovic 

> Is it accepted to not warn at all, as the new default behavior (no option) which would solve both the bug and this issue?

Personally, yes.

If there is a comment in a parameter list, I think someone intentionally did so, therefore the relative comment location should be kept. Removing parentheses of a parameter list moves the relative comment location from "before a parameter" to "before a function". We cannot remove the parentheses along with keeping the "before a parameter", so I consider the parentheses are needed.

**Comment by mdjermanovic** [comment]
Agreed, thanks for clarifying!

**Comment by anikethsaha** [comment]
Is anyone working on this? if not, I can give this a go 👍

**Comment by mdjermanovic** [comment]
> Is anyone working on this? if not, I can work on this

Most likely not, feel free to work on this!
